File: agenkit-java/src/main/java/io/agenkit/budget/CostTracker.java
package io.agenkit.budget;

import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Tracks cumulative LLM usage costs.
 */
public final class CostTracker {

    private final AtomicLong totalInputTokens = new AtomicLong(0);
    private final AtomicLong totalOutputTokens = new AtomicLong(0);
    private final AtomicReference<Double> totalCostUsd = new AtomicReference<>(0.0);
    private final String model;

    public CostTracker(String model) {
        this.model = model;
    }

    public void record(int inputTokens, int outputTokens) {
        totalInputTokens.addAndGet(inputTokens);
        totalOutputTokens.addAndGet(outputTokens);
        double cost = ModelPricing.calculateCost(model, inputTokens, outputTokens);
        totalCostUsd.updateAndGet(prev -> prev + cost);
    }

    public long getTotalInputTokens() { return totalInputTokens.get(); }
    public long getTotalOutputTokens() { return totalOutputTokens.get(); }
    public double getTotalCostUsd() { return totalCostUsd.get(); }

    public void reset() {
        totalInputTokens.set(0);
        totalOutputTokens.set(0);
        totalCostUsd.set(0.0);
    }
}

File: agenkit-scala/src/main/scala/io/agenkit/budget/CostTracker.scala
package io.agenkit.budget

import java.util.concurrent.atomic.AtomicLong

/** Accumulates token usage and computes estimated USD cost for a given model. */
class CostTracker(
  model: String = "gpt-4o"
):
  private val _totalInputTokens  = new AtomicLong(0)
  private val _totalOutputTokens = new AtomicLong(0)
  private val _requestCount      = new AtomicLong(0)

  def recordUsage(inputTokens: Int, outputTokens: Int): Unit =
    _totalInputTokens.addAndGet(inputTokens)
    _totalOutputTokens.addAndGet(outputTokens)
    _requestCount.incrementAndGet()
    ()

  def totalInputTokens: Long  = _totalInputTokens.get()
  def totalOutputTokens: Long = _totalOutputTokens.get()
  def requestCount: Long      = _requestCount.get()

  def totalCost: Double =
    ModelPricing.estimateCost(model, _totalInputTokens.get().toInt, _totalOutputTokens.get().toInt)

  def reset(): Unit =
    _totalInputTokens.set(0)
    _totalOutputTokens.set(0)
    _requestCount.set(0)

  def summary(): Map[String, Any] = Map(
    "model"         -> model,
    "requests"      -> requestCount,
    "inputTokens"   -> totalInputTokens,
    "outputTokens"  -> totalOutputTokens,
    "estimatedCost" -> totalCost
  )
